core/src/main/java/org/lflang/util/FlexPRETUtil.java

```diff
@@ -0,0 +1,81 @@
+package org.lflang.util;
+
+import java.io.IOException;
+import java.util.Collections;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.lflang.FileConfig;
+import org.lflang.MessageReporter;
+import org.lflang.generator.GeneratorCommandFactory;
+import org.lflang.generator.LFGeneratorContext;
+import org.lflang.target.TargetConfig;
+import org.lflang.target.property.PlatformProperty;
+
+public class FlexPRETUtil {
+  private LFGeneratorContext context;
+  private GeneratorCommandFactory commandFactory;
+  private MessageReporter messageReporter;
+
+  public FlexPRETUtil(
+      LFGeneratorContext context,
+      GeneratorCommandFactory commandFactory,
+      MessageReporter messageReporter) {
+    this.context = context;
+    this.commandFactory = commandFactory;
+    this.messageReporter = messageReporter;
+  }
+
+  public void flashTarget(FileConfig fileConfig, TargetConfig config) {
+    var binPath = fileConfig.binPath;
+    var exeName = fileConfig.name;
+
+    /** FlexPRET's SDK generates a runnable bash script. We just need to run it. */
+    var cmd = binPath + "/" + exeName;
+    var args = Collections.<String>emptyList();
+
+    try {
+      LFCommand flash = commandFactory.createCommand(cmd, args);
+
+      /**
+       * Check if environment variable FP_SDK_FPGA_INTERFACE_PROGRAM is set. If it is set, we should
+       * let the user know it is not supported by LF.
+       *
+       * <p>We cannot support it because FP_SDK_FPGA_INTERFACE_PROGRAM will lead to a [Y/n] prompt,
+       * and the LFCommands do not have a way of getting the prompt response.
+       *
+       * <p>When FP_SDK_FPGA_INTERFACE_PROGRAM is not set, there is no prompt.
+       */
+      var envInterfaceProgram = System.getenv("FP_SDK_FPGA_INTERFACE_PROGRAM");
+
+      /**
+       * FP_SDK_FPGA_INTERFACE_PROGRAM is only relevant when the user has specified board = fpga and
+       * flash = true. So only print the info message if that is the case.
+       */
+      var platform = config.get(PlatformProperty.INSTANCE);
+      var interfaceProgramRelevant =
+          platform.board().value() != null
+              && platform.board().value().equals("fpga")
+              && platform.flash().value() == true;
+
+      if (interfaceProgramRelevant && envInterfaceProgram != null) {
+        messageReporter
+            .nowhere()
+            .info("FP_SDK_FPGA_INTERFACE_PROGRAM not supported by LF. It will have no effect.");
+      }
+
+      if (flash != null) {
+        /** Set the variable to nothing for the flash command, which will avoid the [Y/n] prompt */
```

Comment:
THis was also not clear. Setting FP_SDK_FPGA_INTERFACE_PROGRAM to an empty string avoids a [Y/n] prompt? From where?

Reply:
That is from this part of `fpga-app.sh.in`:

```
# Exits if not set
if [[ -z "${FP_SDK_FPGA_INTERFACE_PROGRAM}" ]]; then
    exit 0
fi
# Source the user selected program to interface with the FPGA, e.g., `picocom.source`
# This sets the `interface_cmd` variable
source "$interface_cmd_sourcefile"

# Print out info on how to interface with software
echo "To interface with the FPGA, run:"
echo "$interface_cmd"

# Check if `xclip` exists
if ! command -v xclip > /dev/null 2 >& 1
then
    # Let the user know they can install `xclip` to automatically get command
    # copied to clipboard
    echo "Install \`xclip\` to automatically copy command to clipboard."
else
    # We ask to copy command; do not want to overwrite clipboard if user
    # has something important there
    echo "Copy command to clipboard? [Y/n]"
    read answer

    # Default response is yes
    if [ ! "$answer" != "${answer#[Nn]}" ] ; then 
        # Copy command to clipboard
        echo "$interface_cmd" | xclip -selection clipboard
        echo "Command copied to clipboard. Use CTRL + V to paste."
    fi
fi
```

Reply:
But yes, I can see how it is not at all obvious from the Java code...

Reply:
So if FP_SDK_FPGA_INTERFACE_PROGRAM is empty or not set, it returns early and then what? Is it flashed with some default program? If so, couldn't we just require the user to install that program to avoid this logic?

Reply:
Maybe it's better to just add the entire file here. As you can see, flashing is done independently of whether `FP_SDK_FPGA_INTERFACE_PROGRAM` is set or not (with `send_uart.py`). The `FP_SDK_FPGA_INTERFACE_PROGRAM` refers to what program the user wants to capture standard output with - e.g., `picocom`, `putty`, or some other serial listener program. 

```
# !/bin/bash
#
# This file is auto generated by CMake and should not be modified. 
#
# This script will attempt to upload your compiled software to an FPGA, 
# assuming correct hardware setup and that it runs the bootloader found in 
# `$FP_SDK_PATH/bootloader`.
#
# You got this script because the CMake variable `TARGET` was set to
# `fpga`. Either specify `emulator` or nothing to get a script that runs
# the emulator instead.

set -e

if [[ -z "${FP_PATH}" ]]; then
    echo "Environment variable FP_PATH not set!"
    exit 1
fi

if [[ -z "${FP_SDK_PATH}" ]]; then
    echo "Environment variable FP_SDK_PATH not set!"
    exit 1
fi

flashdev=@FP_FLASH_DEVICE@
baudrate=@UART_BAUDRATE@
imem_path=@IMEM_LOCATION@/@IMEM_NAME@
imem_serialized_path=/tmp/@IMEM_NAME@.serialized
interface_cmd_sourcefile="${FP_SDK_PATH}/cmake/utils/fpga-interface/@FP_INTERFACE_PROGRAM@.source"

echo "Serializing app and placing its result in /tmp"
python3 ${FP_PATH}/scripts/serialize_app.py ${imem_path} ${imem_serialized_path}

echo "Flashing serialized app on device ${flashdev} with baudrate ${baudrate}"

# Script takes usb_path, baudrate, and file
python3 ${FP_PATH}/scripts/send_uart.py ${flashdev} ${baudrate} ${imem_serialized_path}

if [[ -z "${FP_SDK_FPGA_INTERFACE_PROGRAM}" ]]; then
    exit 0
fi
# Source the user selected program to interface with the FPGA, e.g., `picocom.source`
# This sets the `interface_cmd` variable
source "$interface_cmd_sourcefile"

# Print out info on how to interface with software
echo "To interface with the FPGA, run:"
echo "$interface_cmd"

# Check if `xclip` exists
if ! command -v xclip > /dev/null 2 >& 1
then
    # Let the user know they can install `xclip` to automatically get command
    # copied to clipboard
    echo "Install \`xclip\` to automatically copy command to clipboard."
else
    # We ask to copy command; do not want to overwrite clipboard if user
    # has something important there
    echo "Copy command to clipboard? [Y/n]"
    read answer

    # Default response is yes
    if [ ! "$answer" != "${answer#[Nn]}" ] ; then 
        # Copy command to clipboard
        echo "$interface_cmd" | xclip -selection clipboard
        echo "Command copied to clipboard. Use CTRL + V to paste."
    fi
fi

```